src/module_wrap.cc

```diff
@@ -494,23 +493,109 @@ Maybe<uv_file> CheckFile(const URL& search,
   return Just(fd);
 }
 
+PackageConfig emptyPackage = { false, false, "", false };
+std::unordered_map<std::string, PackageConfig> pjson_cache_;
+
+PackageConfig& GetPackageConfig(Environment* env, const std::string path) {
+  auto existing = pjson_cache_.find(path);
+  if (existing != pjson_cache_.end()) {
+    return existing->second;
+  }
+  Maybe<uv_file> check = CheckFile(path, LEAVE_OPEN_AFTER_CHECK);
+  if (check.IsNothing()) {
+    return pjson_cache_[path] = emptyPackage;
+  }
+
+  Isolate* isolate = env->isolate();
+  Local<Context> context = isolate->GetCurrentContext();
+  std::string pkg_src = ReadFile(check.FromJust());
+  uv_fs_t fs_req;
+  uv_fs_close(nullptr, &fs_req, check.FromJust(), nullptr);
+  uv_fs_req_cleanup(&fs_req);
+
+  // It's not okay for the called of this method to not be able to tell
+  // whether an exception is pending or not.
+  TryCatch try_catch(isolate);
+
+  Local<String> src;
+  if (!String::NewFromUtf8(isolate,
+                           pkg_src.c_str(),
+                           v8::NewStringType::kNormal,
+                           pkg_src.length()).ToLocal(&src)) {
+    return pjson_cache_[path] = emptyPackage;
+  }
+
+  Local<Value> pkg_json;
+  if (!JSON::Parse(context, src).ToLocal(&pkg_json) || !pkg_json->IsObject())
+    return (pjson_cache_[path] = emptyPackage);
+  Local<Value> pkg_main;
+  bool has_main = false;
+  std::string main_std;
+  if (pkg_json.As<Object>()->Get(context, env->main_string())
+                              .ToLocal(&pkg_main) && pkg_main->IsString()) {
+    has_main = true;
+    Utf8Value main_utf8(isolate, pkg_main.As<String>());
+    main_std = std::string(*main_utf8, main_utf8.length());
+  }
+
+  bool esm = false;
+  Local<Value> pkg_mode;
+  std::string pkg_mode_std;
+  if (pkg_json.As<Object>()->Get(context, env->mode_string())
+                              .ToLocal(&pkg_mode) && pkg_mode->IsString()) {
+    Utf8Value pkg_mode_utf8(isolate, pkg_mode.As<String>());
+    pkg_mode_std = std::string(*pkg_mode_utf8, pkg_mode_utf8.length());
+    if (pkg_mode_std == "esm") {
```

Comment:
I wonder if it shouldn't throw if package mode is invalid:
- helps to detect typos
- allows us to add new modes without breaking packages

Reply:
The reason for this is for perfect backwards compatibility - on the obscure off-chance that packages already exist on npm with a "mode" property, these won't suddenly break. If we could guarantee that wasn't the case I'd be all for strict validation though.

Reply:
I checked this in BigQuery:
```sql
#standardSQL
SELECT
  repo_name,
  path,
  content
FROM
  `bigquery-public-data.github_repos.sample_files` AS sample_files
JOIN
  `bigquery-public-data.github_repos.sample_contents` AS sample_contents
ON
  sample_contents.id = sample_files.id
WHERE
  REGEXP_CONTAINS(path,
    r"package\.json$")
  AND REGEXP_CONTAINS(content,
    r"\"mode\"")
```

There are 4 results but they are false positives (found in keywords or not top-level).
I cannot run this query on the full dataset (exceeds my quota).